teps/0004-tekton-integrations-annotations-statuses.md

```diff
@@ -0,0 +1,689 @@
+/ tool will notice and act on these annotations.
+
+**Examples:**
+
+- GitHub
+  ```
+  github.integrations.tekton.dev/owner: wlynch
+  github.integrations.tekton.dev/repo: test
+  github.integrations.tekton.dev/sha: 4d2f673c600a0f08c3f6737bc3f5c6a6a15d62e6
+  ```
+- Webhook
+  ```
+  webhook.integrations.tekton.dev/url: https://example.com:8080
+  ```
+- Slack
+  ```
+  slack.integrations.tekton.dev/url: https://tektoncd.slack.com
+  slack.integrations.tekton.dev/channel: build-cop
+  ```
+
+### Integration Status
+
+Key questions we will need to answer is - did an integration complete
+successfully for my pipeline? if not, why?
+
+In Tekton's
+[dogfooding cluster](https://github.com/tektoncd/plumbing/blob/master/docs/dogfooding.md)
+today, this answer is somewhat obfuscated. When a pipeline is complete, the
+cluster kicks off a cloud event containing information about the PR to update.
+This is picked up by another Trigger, which then updates the PR. It is difficult
+to tell from the initial PipelineRun whether a PR was updated, and if there was
+a problem, why.
+
+To solve this, we should add additional status fields in
+TaskRunStatus/PipelineRunStatus, specifically for integration data. This should
+include information such as:
+
+- What integration the status is for (e.g. `github.integrations.tekton.dev`).
+- What object generation the status is for.
+- What was the outcome of the integration for this object (e.g. success,
+  failure).
+- (optional) More details of status outcome.
+
+Examples of what might be stored:
+
+- Integration Success/Failure
+- GitHub CheckRun IDs
+- Tekton Result IDs
+- Integration error responses (unavailable, out of quota, etc.)
+
+This is a generalization of the existing
+[`CloudEventDelivery`](https://pkg.go.dev/github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1?tab=doc#CloudEventDelivery)
+status data that is stored in Run statuses today. [See below](#notifications)
+For more on how this to the ongoing notifications effort.
+
+### User Stories (optional)
+
+<!--
+Detail the things that people will be able to do if this TEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system.  The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Error updates on Pipeline failures
+
+By making integration code independent of Pipeline execution, this enables us to
+respond to external integrations even in cases where the Pipeline/Task fails to
+start, as long as the Pipeline/TaskRun has been accepted by Tekton.
+
+This is particularly useful for validation / dereferencing errors that cannot be
+caught at the time of initial validation.
+
+#### Why didn't this Run update my PR?
+
+With integration statuses, we have a place to annotate data alongside the
+pipeline run itself. This gives us an easy mechanism for users to do the
+following:
+
+1. Push a commit, notice pull request wasn't updated for a run.
```

Comment:
Independently of this (but maybe related), an integration should *ensure* to do the proper thing whenever the `*Run` is created. e.g. for GitHub, as soon as the `*Run` is created, the checks that this `*Run` provides should be reseted (to `pending`). *But then, the question is how do you know which `checks` this `*Run` provides ?*